### PR TITLE
feat: cast between `Intervals`

### LINKED ISSUE #4181: Cast between `Intervals`
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
`arrow-rs` now does not support cast between `Intervals`.
This can cause difficulties when working with dissimilar intervals.

The example of code:
```
use arrow::array::*;
use arrow_cast::cast;
use arrow_schema::DataType;
use std::sync::Arc;
use arrow::datatypes::{IntervalUnit, IntervalYearMonthType};

fn main() {
    let a = IntervalYearMonthArray::from(vec![
        IntervalYearMonthType::make_value(1, 1),
        IntervalYearMonthType::make_value(2, 2),
        IntervalYearMonthType::make_value(3, 3),
    ]);
    let array = Arc::new(a) as ArrayRef;
    let b = cast(&array, &DataType::Interval(IntervalUnit::MonthDayNano)).unwrap();
    let c = b.as_any().downcast_ref::<IntervalMonthDayNanoArray>().unwrap();
    println!("{:#?}", c); // Casting from Interval(YearMonth) to Interval(MonthDayNano) not supported
}
```

**Describe the solution you'd like**
Add the ability for the function to process the following inputs:
`Interval(IntervalUnit::YearMonth)` -> `Interval(IntervalUnit::MonthDayNano)`
`Interval(IntervalUnit::DayTime)` -> `Interval(IntervalUnit::MonthDayNano)`

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by izveigor**
# Which issue does this PR close?

Closes #4181

# Rationale for this change
 

# What changes are included in this PR?

# Are there any user-facing changes?
Yes

### CODE REVIEW

File: arrow-cast/src/cast.rs [hunk 1]
```diff
@@ -460,6 +462,56 @@ where
...
+) -> Result<ArrayRef, ArrowError> {
+    let array = array
+        .as_any()
+        .downcast_ref::<PrimitiveArray<IntervalYearMonthType>>()
+        .ok_or_else(|| {
+            ArrowError::ComputeError(
+                "Internal Error: Cannot cast interval to IntervalArray of expected type"
+                    .to_string(),
+            )
+        })?;
```
Comment (tustvold) [inline arrow-cast/src/cast.rs:478]:
```suggestion
    let array = array.as_primitive::<IntervalYearMonthType>();
```

File: arrow-cast/src/cast.rs [hunk 2]
```diff
@@ -460,6 +462,56 @@ where
...
+) -> Result<ArrayRef, ArrowError> {
+    let array = array
+        .as_any()
+        .downcast_ref::<PrimitiveArray<IntervalDayTimeType>>()
+        .ok_or_else(|| {
+            ArrowError::ComputeError(
+                "Internal Error: Cannot cast interval to IntervalArray of expected type"
+                    .to_string(),
+            )
+        })?;
```
Comment (tustvold) [inline arrow-cast/src/cast.rs:501]:
```suggestion
        let array = array.as_primitive::<IntervalDayTimeType>();
```

File: arrow-cast/src/cast.rs [hunk 3]
```diff
@@ -460,6 +462,56 @@ where
...
+
+    let iter = array.iter().map(|v| {
```
Comment (tustvold) [inline arrow-cast/src/cast.rs:503]: I think this needs to perform checked conversion, and can make use of unary_opt and try_unary to achieve this based on the cast options

File: arrow-cast/src/cast.rs [hunk 4]
```diff
@@ -8761,4 +8819,98 @@ mod tests {
...
+            cast_options,
+        )?;
+        casted_array
+            .as_any()
+            .downcast_ref::<IntervalMonthDayNanoArray>()
+            .ok_or_else(|| {
+                ArrowError::ComputeError(
+                    "Failed to downcast to IntervalMonthDayNanoArray".to_string(),
+                )
+            })
```
Comment (tustvold) [inline arrow-cast/src/cast.rs:8889]:
```suggestion
        Ok(casted_array.as_primitive::<IntervalMonthDataNanoType>().clone())
```

File: arrow-cast/src/cast.rs [hunk 5]
```diff
@@ -8761,4 +8819,98 @@ mod tests {
...
+            .cloned()
```
Comment (tustvold) [inline arrow-cast/src/cast.rs:8843]:
```suggestion
        Ok(casted_array.as_primitive::<IntervalMonthDayNanoType>().clone())
```

**Review by tustvold — COMMENTED** [review]
I think this needs some review of the overflow behaviour

File: arrow-cast/src/cast.rs [hunk 6]
```diff
@@ -409,6 +413,125 @@ where
...
+    if cast_options.safe {
+        Ok(Arc::new(array.unary_opt::<_, IntervalMonthDayNanoType>(
+            |v| {
+                let months = IntervalYearMonthType::to_months(v);
+                Some(IntervalMonthDayNanoType::make_value(months, 0, 0))
+            },
+        )))
+    } else {
+        Ok(Arc::new(
+            array.try_unary::<_, IntervalMonthDayNanoType, ArrowError>(|v| {
```
[the remark this answers is not in the document]
Comment (tustvold) [reply arrow-cast/src/cast.rs:432]: If the method is infallible you can just use `unary`

File: arrow-cast/src/cast.rs [hunk 7]
```diff
@@ -409,6 +413,125 @@ where
...
+    let mul = 10_i32.pow(6);
+
+    if cast_options.safe {
+        Ok(Arc::new(array.unary_opt::<_, IntervalMonthDayNanoType>(
+            |v| {
+                let (days, ms) = IntervalDayTimeType::to_parts(v);
+                Some(IntervalMonthDayNanoType::make_value(
+                    0,
+                    days,
+                    (ms * mul).into(),
```
Comment (tustvold) [inline arrow-cast/src/cast.rs:455]:
```suggestion
                    ms as i64 * mil,
```

File: arrow-cast/src/cast.rs [hunk 8]
```diff
@@ -409,6 +413,125 @@ where
...
+    }
+}
+
+/// Cast the array from interval day time to month day nano
+fn cast_interval_day_time_to_interval_month_day_nano(
+    array: &dyn Array,
+    cast_options: &CastOptions,
+) -> Result<ArrayRef, ArrowError> {
+    let array = array.as_primitive::<IntervalDayTimeType>();
+    let mul = 10_i32.pow(6);
```
Comment (tustvold) [inline arrow-cast/src/cast.rs:446]:
```suggestion
    let mul = 1_000_000;
```

File: arrow-cast/src/cast.rs [hunk 9]
```diff
@@ -409,6 +413,125 @@ where
...
+
+    if cast_options.safe {
```
Comment (tustvold) [inline arrow-cast/src/cast.rs:448]: It occurs to me that this actually cannot overflow, and so can just use `unary`.

It is performing a widening multiplication of i32 milliseconds to i64 nanoseconds

File: arrow-cast/src/cast.rs [hunk 10]
```diff
@@ -409,6 +413,125 @@ where
...
+) -> Result<ArrayRef, ArrowError> {
+    let array = array.as_primitive::<IntervalMonthDayNanoType>();
+    let mul = 10_i64.pow(6);
+
+    if cast_options.safe {
+        Ok(Arc::new(array.unary_opt::<_, IntervalDayTimeType>(|v| {
+            let (_, days, nanos) = IntervalMonthDayNanoType::to_parts(v);
+            Some(IntervalDayTimeType::make_value(
+                days,
+                (nanos / mul).try_into().unwrap(),
```
Comment (tustvold) [inline arrow-cast/src/cast.rs:510]:
```suggestion
                (nanos / mul).try_into().ok()?,
```

File: arrow-cast/src/cast.rs [hunk 11]
```diff
@@ -409,6 +413,125 @@ where
...
+fn cast_interval_month_day_nano_to_interval_day_time(
+    array: &dyn Array,
+    cast_options: &CastOptions,
+) -> Result<ArrayRef, ArrowError> {
+    let array = array.as_primitive::<IntervalMonthDayNanoType>();
+    let mul = 10_i64.pow(6);
+
+    if cast_options.safe {
+        Ok(Arc::new(array.unary_opt::<_, IntervalDayTimeType>(|v| {
+            let (_, days, nanos) = IntervalMonthDayNanoType::to_parts(v);
```
Comment (tustvold) [inline arrow-cast/src/cast.rs:507]: This appears to discard the months component?

File: arrow-cast/src/cast.rs [hunk 12]
```diff
@@ -409,6 +413,125 @@ where
...
+        ))
+    }
+}
+
+/// Cast the array from interval month day nano to year month
+fn cast_interval_month_day_nano_to_interval_year_month(
+    array: &dyn Array,
+    cast_options: &CastOptions,
+) -> Result<ArrayRef, ArrowError> {
+    let array = array.as_primitive::<IntervalMonthDayNanoType>();
```
Comment (tustvold) [inline arrow-cast/src/cast.rs:480]: I think people would find it surprising that this discards any days or nanoseconds. I suspect we either need to error if they are anything other than 0, or make an approximation about the number of hours in a day and the number of days in a month...

**Review by tustvold — COMMENTED** [review]
I left some comments, thinking about this further I think the only well-formed interval casts are from a smaller representation to a larger representation.

I'm not sure there is a "correct" way to go from `IntervalUnit::MonthDayNano` to `IntervalUnit::DayTime`.

However, I think just providing the infallible "widening" casts, is good enough, and will be sufficient for most use-cases. At the very least they seem like a good place to start

**Comment by alamb** [comment]
I wonder if this PR will make the next release (so we can use it in datafusion) 🤔

**Review by tustvold — APPROVED** [review]

**Comment by tustvold** [comment]
Thank you

**Pull request merged**